platforms/platforms.go

```diff
@@ -0,0 +1,129 @@
+package platforms
+
+import (
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/containerd/containerd/errdefs"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type platformKey struct {
+	os      string
+	arch    string
+	variant string
+}
+
+var (
+	selectorRe = regexp.MustCompile(`^[A-Za-z0-9_-]$`)
+)
+
+// ParseSelector parses the platform selector syntax into a platform
+// declaration.
+//
+// Platform selectors are in the format <os|arch>[/<arch>[/<variant>]]. The minimum
+// required information for a platform selector is the operating system.
+//
+// The minimum required information for a platform selector is the operating
+// system or architecture. If there is only a single string (no slashes(, the
+// value will be matched against the known set of operating systems, then fall
+// back to the known set of architectures. The missing component will be
+// inferred based on the local environment.
+func Parse(selector string) (specs.Platform, error) {
+	if strings.Contains(selector, "*") {
+		// TODO(stevvooe): need to work out exact wildcard handling
+		return specs.Platform{}, errors.Wrapf(errdefs.ErrInvalidArgument, "%q: wildcards not yet supported", selector)
+	}
+
+	if selectorRe.MatchString(selector) {
```

Comment:
Looks like this is not matching most due to an error in the regular expression (will only match a single character), but this should check that it does _not_ match, not that it _does_ match, so add a `!` before the condition.